src/cryojax/data/_particle_data/relion.py

```
@@ -0,0 +1,1597 @@
+        )
+        # Properties for loading
+        self._loads_metadata = loads_metadata
+        self._broadcasts_optics_group = broadcasts_optics_group
+        self._loads_envelope = loads_envelope
+        # Properties for writing
+        self._updates_optics_group = updates_optics_group
+
+    @override
+    def __getitem__(
+        self, index: int | slice | Int[np.ndarray, ""] | Int[np.ndarray, " _"]
+    ) -> ParticleParameterInfo:
+        # Validate index
+        n_rows = self.starfile_data["particles"].shape[0]
+        _validate_dataset_index(type(self), index, n_rows)
+        # ... read particle data at the requested indices
+        particle_data = self.starfile_data["particles"]
+        particle_data_at_index = particle_data.iloc[index]
+        # ... read optics group given the particle data
+        optics_data = self.starfile_data["optics"]
+        optics_group = _get_optics_group_from_particle_data(
+            particle_data_at_index, optics_data
+        )
+        # Load the image stack and STAR file parameters
+        instrument_config, transfer_theory, pose = _make_pytrees_from_starfile(
+            particle_data_at_index,
+            optics_group,
+            self.broadcasts_optics_group,
+            self.loads_envelope,
+            self._make_config_fn,
+        )
+        # ... convert to dataframe for serialization
+        if isinstance(particle_data_at_index, pd.Series):
+            particle_data_at_index = particle_data_at_index.to_frame().T
+        metadata = particle_data_at_index.to_dict() if self.loads_metadata else {}
+        return ParticleParameterInfo(
+            instrument_config=instrument_config,
+            pose=pose,
+            transfer_theory=transfer_theory,
+            metadata=metadata,
+        )
+
+    @override
+    def __len__(self) -> int:
+        return len(self.starfile_data["particles"])
+
+    @override
+    def __setitem__(
```

Assuming the current behavior is the desired behavior:

It should be clear that `__setitem__` should only be used to update parameters. In addition, it should raise an error when indices that do not exist are "set". For example, I generated a new parameters file, then tried `param_file[0:10] = parameters`, and it ran without issue, but it did not really update the parameters in the file, the file was still empty. This can be solved by using append, but it should either work with setitem or it should raise an error.

---

This should raise an error—in this case the user has to use append, since these indices don’t yet exist (similar to list / array behavior). The _validate_dataset_index function should raise the error, we can see why that’s not doing so

---

Solved!